### PR TITLE
Fix: Lambda `servicecachehealth` times out (#5467, #5622)

### LINKED ISSUE #5467: Lambda `servicecachehealth` times out
… when requesting `https://service.prod.anvil.gi.ucsc.edu/index/files?size=1` to build the health object.

[Logs Insights](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-11T15*3a28*3a52.770Z~start~'2023-08-11T15*3a28*3a29.903Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20like*20*2ff3feb75c-6d88-4a66-9534-d82309b5ffd1*2f*0a*7c*20sort*20*40timestamp*20desc*0a*7c*20limit*2020~queryId~'e3b8339c4fddfef7-be1a72f7-4f589d-67d32e9-b695460acc1cce4cddd7a46~source~(~'arn*3aaws*3alogs*3aus-east-1*3a465330168186*3alog-group*3a*2faws*2flambda*2fazul-service-anvilprod-servicecachehealth~'arn*3aaws*3alogs*3aus-east-1*3a465330168186*3alog-group*3a*2faws*2fapigateway*2fazul-service-anvilprod)))
```
2023-08-11T15:28:48.006Z f3feb75c-6d88-4a66-9534-d82309b5ffd1 Task timed out after 10.02 seconds
```
EDIT: (@achave11-ucsc)

Better logging has now allowed for the detection of an S3 PUT operation timing out. It's somewhat surprising since the object isn't of significant size.

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-service-prod-servicecachehealth    
start-time: 2024-01-04T18:34:57.841Z    
end-time: 2024-01-05T04:01:43.597Z    
query-string:
  ```
  fields @timestamp
| filter @message like /0d162516/
| filter @message not like /END/
| parse @message '05fb54b90376*' as msg
| parse @log 'lambda/*' as lambda
| sort @timestamp asc
| limit 20
  ```
---
| @timestamp | msg | lambda |
| --- | --- | --- |
| 2024-01-04 23:53:29.896 | Version: $LATEST | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:29.896 | azul.health Getting health property 'api_endpoints' | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:29.897 | azul.health Making HEAD request to https://service.azul.data.humancellatlas.org/index/files?size=1 | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:34.963 | azul.health Got 200 response after 5.065s from HEAD request to https://service.azul.data.humancellatlas.org/index/files?size=1 | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:34.964 | azul.health Getting health property 'elasticsearch' | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:34.964 | elasticsearch Making HEAD request to https://vpc-azul-index-prod-mjluewlvxikvhillefxotw2tbm.us-east-1.es.amazonaws.com:443/ | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:34.964 | elasticsearch … without request body | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:35.001 | elasticsearch Got 200 response after 0.037s from HEAD to https://vpc-azul-index-prod-mjluewlvxikvhillefxotw2tbm.us-east-1.es.amazonaws.com:443/ | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:35.001 | elasticsearch … with response body '' | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:35.002 | azul.boto3 s3.PutObject: Making PUT request to https://s3.amazonaws.com/edu-ucsc-gi-platform-hca-prod-storage-prod.us-east-1/health/service | azul-service-prod-servicecachehealth |
| 2024-01-04 23:53:35.002 | azul.boto3 s3.PutObject: … with nonprintable body (<class '_io.BytesIO'>) | azul-service-prod-servicecachehealth |
| 2024-01-04 23:54:08.983 | Task timed out after 39.09 seconds | azul-service-prod-servicecachehealth |
| 2024-01-04 23:54:08.983 | Duration: 39087.51 ms Billed Duration: 39000 ms Memory Size: 128 MB Max Memory Used: 121 MB | azul-service-prod-servicecachehealth |
---

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by achave11-ucsc**
Connected issues: #5467, #5622

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>
- [x] All connected issues are resolved partially <sub>or this PR does not have the `partial` label</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading deployments)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading deployments</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from system administrator
- [x] PR is assigned to system administrator

### System administrator (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvilprod`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only include `p` if the PR is labeled `partial`</sub>
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Deployed `dev.gitlab`
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvildev.gitlab`
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvilprod.gitlab`
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvilprod`</sub>
- [x] Considered deindexing individual sources in `dev` <sub>or this PR does not merely remove sources from existing catalogs in `dev`</sub>
- [x] Considered deindexing individual sources in `anvildev` <sub>or this PR does not merely remove sources from existing catalogs in `anvildev`</sub>
- [x] Considered deindexing individual sources in `anvilprod` <sub>or this PR does not merely remove sources from existing catalogs in `anvilprod`</sub>
- [x] Considered indexing individual sources in `dev` <sub>or this PR does not merely add sources to existing catalogs in `dev`</sub>
- [x] Considered indexing individual sources in `anvildev` <sub>or this PR does not merely add sources to existing catalogs in `anvildev`</sub>
- [x] Considered indexing individual sources in `anvilprod` <sub>or this PR does not merely add sources to existing catalogs in `anvilprod`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing `anvilprod`</sub>

### Operator

- [x] Added CL items to upcoming promotion PR
> - [ ] Deployed `prod.gitlab`
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 1]
```diff
@@ -1017,6 +1017,15 @@ def qq(*words):
             }
             for zone in range(num_zones)
         },
+        'aws_vpc_endpoint': {
+            f'{service}_gateway': {
+                'service_name': f'com.amazonaws.{config.region}.{service}',
+                'vpc_endpoint_type': 'Gateway',
+                'vpc_id': '${aws_vpc.gitlab.id}',
+                'route_table_ids': [f'${{aws_route_table.gitlab_{i}.id}}' for i in range(num_zones)],
```
Comment (nadove-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]: I believe we prefer using `%` for string interpolation in conjunction with Terraform's `{}` syntax

```suggestion
                'route_table_ids': ['${aws_route_table.gitlab_%d.id}' % i for i in range(num_zones)],
```

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 2]
```diff
@@ -1017,6 +1017,15 @@ def qq(*words):
...
+            }
+            for service in ('dynamodb', 's3')
```
Comment (nadove-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]:
```suggestion
            for service in ['dynamodb', 's3']
```

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]

**Review by nadove-ucsc — DISMISSED** [review]

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 3]
```diff
@@ -1017,6 +1017,15 @@ def qq(*words):
             }
             for zone in range(num_zones)
         },
+        'aws_vpc_endpoint': {
+            f'{service}_gateway': {
+                'service_name': f'com.amazonaws.{config.region}.{service}',
+                'vpc_endpoint_type': 'Gateway',
+                'vpc_id': '${aws_vpc.gitlab.id}',
+                'route_table_ids': ['${aws_route_table.gitlab_%d.id}' % i for i in range(num_zones)],
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]: L

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 4]
```diff
@@ -1017,6 +1017,15 @@ def qq(*words):
             }
             for zone in range(num_zones)
         },
+        'aws_vpc_endpoint': {
+            f'{service}_gateway': {
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]:
```suggestion
            f'gitlab_{service}': {
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Despite this being `no sandbox` please don't skip pushing the merge commit.

**Pull request merged**